extensions/pkg/controller/healthcheck/worker/nodes.go

```diff
@@ -0,0 +1,194 @@
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	gardencorev1beta1helper "github.com/gardener/gardener/pkg/apis/core/v1beta1/helper"
+
+	machinev1alpha1 "github.com/gardener/machine-controller-manager/pkg/apis/machine/v1alpha1"
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// DefaultHealthChecker all the information for the Worker HealthCheck.
+// This check assumes that the MachineControllerManager (https://github.com/gardener/machine-controller-manager) has been
+// deployed by the Worker extension controller.
+type DefaultHealthChecker struct {
+	logger logr.Logger
+	// Needs to be set by actuator before calling the Check function
+	seedClient client.Client
+	// make sure shoot client is instantiated
+	shootClient client.Client
+	// scaleUpProgressingThreshold is the progressing threshold when the health check detects a scale-up situation.
+	scaleUpProgressingThreshold *time.Duration
+	// scaleDownProgressingThreshold is the progressing threshold when the health check detects a scale-down situation.
+	scaleDownProgressingThreshold *time.Duration
+}
+
+// NewNodesChecker is a health check function which performs certain checks about the nodes registered in the cluster.
+// It implements the healthcheck.HealthCheck interface.
+func NewNodesChecker() *DefaultHealthChecker {
+	scaleUpProgressingThreshold := 5 * time.Minute
+	scaleDownProgressingThreshold := 15 * time.Minute
+
+	return &DefaultHealthChecker{
+		scaleUpProgressingThreshold:   &scaleUpProgressingThreshold,
+		scaleDownProgressingThreshold: &scaleDownProgressingThreshold,
+	}
+}
+
+// WithScaleUpProgressingThreshold sets the scaleUpProgressingThreshold property.
+func (h *DefaultHealthChecker) WithScaleUpProgressingThreshold(d time.Duration) *DefaultHealthChecker {
+	h.scaleUpProgressingThreshold = &d
+	return h
+}
+
+// WithScaleDownProgressingThreshold sets the scaleDownProgressingThreshold property.
+func (h *DefaultHealthChecker) WithScaleDownProgressingThreshold(d time.Duration) *DefaultHealthChecker {
+	h.scaleDownProgressingThreshold = &d
+	return h
+}
+
+// InjectSeedClient injects the seed client.
+func (h *DefaultHealthChecker) InjectSeedClient(seedClient client.Client) {
+	h.seedClient = seedClient
+}
+
+// InjectShootClient injects the shoot client.
+func (h *DefaultHealthChecker) InjectShootClient(shootClient client.Client) {
+	h.shootClient = shootClient
+}
+
+// SetLoggerSuffix injects the logger.
+func (h *DefaultHealthChecker) SetLoggerSuffix(provider, extension string) {
+	h.logger = log.Log.WithName(fmt.Sprintf("%s-%s-healthcheck-nodes", provider, extension))
+}
+
+// DeepCopy clones the healthCheck struct by making a copy and returning the pointer to that new copy.
+func (h *DefaultHealthChecker) DeepCopy() healthcheck.HealthCheck {
```

Comment:
This is actually not deep copying the healthcheck, although the function name suggests it... Can you rename the interface func to `Clone` or something?

Reply:
This is not my code, please get in touch with @danielfoehrKn. It cannot be renamed because it implements the `healthcheck.HealthCheck` interface which expects it to be called `DeepCopy`.

Reply:
Ok, sure.